Sources/NIOCore/AsyncSequences/NIOThrowingAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,503 @@
+            self.storage.yield(sequence)
+        }
+
+        /// Yields a new elements to the ``NIOThrowingAsyncSequenceProducer``.
+        ///
+        /// If there is an ``NIOThrowingAsyncSequenceProducer/AsyncIterator`` awaiting the next element, it will get resumed right away.
+        /// Otherwise, the element will get buffered.
+        ///
+        /// If the ``NIOThrowingAsyncSequenceProducer`` is terminated this will drop the elements
+        /// and return ``YieldResult/dropped``.
+        ///
+        /// This can be called more than once and returns to the caller immediately
+        /// without blocking for any awaiting consumption from the iteration.
+        ///
+        /// - Parameter element: The element to yield.
+        /// - Returns: A ``NIOThrowingAsyncSequenceProducer/Source/YieldResult`` that indicates if the yield was successful
+        /// and if more elements should be produced.
+        @inlinable
+        public func yield(_ element: Element) -> YieldResult {
+            self.yield(contentsOf: CollectionOfOne(element))
+        }
+
+        /// Finishes the sequence.
+        ///
+        /// Calling this function signals the sequence that there won't be any subsequent elements yielded.
+        ///
+        /// If there are still buffered elements and there is an ``NIOThrowingAsyncSequenceProducer/AsyncIterator`` consuming the sequence,
+        /// then termination of the sequence only happens once all elements have been consumed by the ``NIOThrowingAsyncSequenceProducer/AsyncIterator``.
+        /// Otherwise, the buffered elements will be dropped.
+        ///
+        /// - Note: Calling this function more than once has no effect.
+        @inlinable
+        public func finish() {
+            self.storage.finish(nil)
+        }
+
+        /// Finishes the sequence with the given `Failure`.
+        ///
+        /// Calling this function signals the sequence that there won't be any subsequent elements yielded.
+        ///
+        /// If there are still buffered elements and there is an ``NIOThrowingAsyncSequenceProducer/AsyncIterator`` consuming the sequence,
+        /// then termination of the sequence only happens once all elements have been consumed by the ``NIOThrowingAsyncSequenceProducer/AsyncIterator``.
+        /// Otherwise, the buffered elements will be dropped.
+        ///
+        /// - Note: Calling this function more than once has no effect.
+        /// - Parameter failure: The failure why the sequence finished.
+        @inlinable
+        public func finish(_ failure: Failure) {
+            self.storage.finish(failure)
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOThrowingAsyncSequenceProducer {
+    /// This is the underlying storage of the sequence. The goal of this is to synchronize the access to all state.
+    @usableFromInline
+    /* fileprivate */ internal final class Storage: @unchecked Sendable {
```

Comment:
@Lukasa This type could also be unified between the two implementations. I didn't do that yet because the return type of `yield() -> Source.YieldResult` is currently depending on the type. There are two ways forward:

1. Not unify this type
2. Move the `YieldResult` outside of the generic types and unify the `Storage` type.

Let me know what you prefer!